network/network-address/src/encrypted.rs

```diff
@@ -0,0 +1,482 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::RawNetworkAddress;
+use aes_gcm::{
+    aead::{generic_array::GenericArray, AeadInPlace, NewAead},
+    Aes256Gcm,
+};
+use libra_crypto::{
+    compat::Sha3_256,
+    hash::HashValue,
+    hkdf::{Hkdf, HkdfError},
+};
+use move_core_types::account_address::AccountAddress;
+use once_cell::sync::Lazy;
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::{convert::TryFrom, mem};
+use thiserror::Error;
+
+/// The length in bytes of the AES-256-GCM authentication tag.
+pub const AES_GCM_TAG_LEN: usize = 16;
+
+/// The length in bytes of the AES-256-GCM nonce.
+pub const AES_GCM_NONCE_LEN: usize = 12;
+
+/// The length in bytes of the root key and derived per-validator account keys.
+pub const KEY_LEN: usize = 32;
+
+/// A root key used to derive per-validator account keys.
+///
+/// This key should be shared amongst the validators for encrypting and decrypting
+/// the validator network addresses. There may be multiple versions of the root
+/// key, disambiguated by their `root_key_version`.
+pub type RootKey = [u8; KEY_LEN];
+pub type RootKeyVersion = u32;
+
+// Constant root key + version so we can push `EncNetworkAddress` everywhere
+// without worrying about getting the key in the right places. these will be
+// test-only soon.
+// TODO(philiphayes): feature gate for testing/fuzzing only
+pub const TEST_ROOT_KEY: RootKey = [0u8; KEY_LEN];
+pub const TEST_ROOT_KEY_VERSION: RootKeyVersion = 0;
+
+/// We salt the HKDF for deriving the account keys to provide application
+/// separation.
+///
+/// Note: modifying this salt is a backwards-incompatible protocol change.
+///
+/// For readers, this hash value is equal to the following hex string:
+/// `"dfc8ffcc7f62ea4e5b9bc41ee7969b44275419ebaad1db27d2a191b6d1db6d13"`
+pub static HKDF_SALT: Lazy<HashValue> =
+    Lazy::new(|| HashValue::sha3_256_of(b"LIBRA_ENCRYPTED_NETWORK_ADDRESS_SALT"));
+
+/// A serialized `EncNetworkAddress`.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct RawEncNetworkAddress(#[serde(with = "serde_bytes")] Vec<u8>);
```

Comment:
why specify serde_bytes?

Reply:
it makes the serialization/deserialization of bytes much more efficient. without the annotation, it will literally ser/de a single byte at a time, whereas with the annotation it will just memcpy the whole chunk in one go